src/hidpp20.c

```diff
@@ -2412,91 +2421,60 @@ hidpp20_onboard_profiles_parse(struct hidpp20_device *device,
-		}
-	}
+					       "profile %d: error in the address: 0x%04x instead of 0x%04x\n",
+					       i + 1,
+					       profiles->profiles[i].address,
+					       HIDPP20_USER_PROFILES_G402 | (i + 1));
 
-	/*
-	 * Even if the crc is wrong, we still fetch all the profiles, so as
-	 * to have the current stored values if some profiles are legitimate.
-	 */
-	 for (i = 0; i < profiles->num_profiles; i++) {
-		if (profiles->profiles[i].address == 0x0000) {
-			switch (dict_address) {
-			case 0x0000:
-				profiles->profiles[i].address = i + 1;
-				break;
-			case 0x0100:
-				if (i >= profiles->num_rom_profiles) {
-					profiles->profiles[i].address = dict_address | profiles->num_rom_profiles;
-				} else {
-					profiles->profiles[i].address = dict_address | (i + 1);
-				}
-				break;
-			default:
-				hidpp_log_error(&device->base,
-						"this should never happen: dict address is 0x%04x\n",
-						dict_address);
-				return -EINVAL;
-			}
+			profiles->profiles[i].enabled = !!d[HIDPP20_PROFILE_DIR_ENABLED];
 		}
-		rc = hidpp20_onboard_profiles_parse_profile(device,
-							    profiles,
-							    i,
-							    check_crc);
-		if (rc == -EAGAIN)
-			crc_valid = false;
-	}
-
-	if (check_crc && !crc_valid)
-		return -EAGAIN;
-
-	return 0;
-}
+	} else {
+		hidpp_log_debug(device, "Profile directory has an invalid CRC... Reading ROM profiles.\n");
 
-int
-hidpp20_onboard_profiles_initialize(struct hidpp20_device *device,
-				    struct hidpp20_profiles *profiles)
-{
-	int rc;
-
-	assert(profiles);
-
-	rc = hidpp20_onboard_profiles_parse(device, profiles, 0x0000, true);
-	if (rc == 0)
-		return profiles->num_profiles;
-
-	if (rc != -EAGAIN) {
-		/* something went wrong */
-		return rc;
+		read_userdata = false;
 	}
 
-	/*
-	 * One of the used sector has a bad crc, the device is using the ROM
-	 * profiles.
-	 */
-
-	hidpp_log_debug(&device->base, "device is using ROM settings\n");
+	for (i = 0; i < profiles->num_profiles; i++) {
+		if (read_userdata) {
+			hidpp_log_debug(device, "Parsing profile %u\n", i);
+			rc = hidpp20_onboard_profiles_parse_profile(device,
+								    profiles,
+								    i,
+								    true);
+
+			/* on fail to read the user profile fallback to the default profile */
+			if (rc)
+				hidpp_log_debug("Profile %u is bad. Falling back to the ROM settings.\n", i);
+			else
+				continue;
+		}
 
-	rc = hidpp20_onboard_profiles_parse(device, profiles, 0x0100, false);
-	if (rc < 0)
-		return rc;
+		/* the number of rom profiles can be different than the number of user profiles
+		   so we if there are not enough rom profiles to populate all the user profiles
+		   we just use the first rom profile */
+		if (i + 1 > profiles->num_rom_profiles)
+			profiles->profiles[i].address = HIDPP20_ROM_PROFILES_G402 + 1;
+		else
+			profiles->profiles[i].address = HIDPP20_ROM_PROFILES_G402 + i + 1;
 
-	hidpp_log_debug(&device->base, "ROM settings loaded\n");
+		rc = hidpp20_onboard_profiles_parse_profile(device,
```

Comment:
The value of rc is now never checked.
We might want to have a `if (rc <0) return rc;` here